src/libcore/alloc.rs

```diff
@@ -900,6 +976,62 @@ pub unsafe trait Alloc {
         }
     }
 
+    /// Attempts to shrink the allocation referenced by `ptr` to fit `new_size`.
+    /// For some `layout` inputs, like arrays, this may include extra storage
+    /// usable for additional data.
+    ///
+    /// If this returns `Ok`, then the allocator has asserted that the
+    /// memory block referenced by `ptr` now fits `new_size`, and
+    /// thus can only be used to carry data of that smaller
+    /// layout. (The allocator is allowed to take advantage of this,
+    /// carving off portions of the block for reuse elsewhere.) The
+    /// truncated contents of the block within the smaller layout are
+    /// unaltered, and ownership of block has not been transferred.
+    ///
+    /// If this returns `Err`, then the memory block is considered to
+    /// still represent the original (larger) `layout`. None of the
+    /// block has been carved off for reuse elsewhere, ownership of
+    /// the memory block has not been transferred, and the contents of
+    /// the memory block are unaltered.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because undefined behavior can result
+    /// if the caller does not ensure all of the following:
+    ///
+    /// * `ptr` must be currently allocated via this allocator,
+    ///
+    /// * `layout` must *fit* the `ptr` (see above); note the
+    ///   `new_size` argument need not fit it,
+    ///
+    /// * `new_size` must not be greater than `layout.size()`
+    ///   (and must be greater than zero),
+    ///
+    /// # Errors
+    ///
+    /// Returns `Err(CannotReallocInPlace)` when the allocator is
+    /// unable to assert that the memory block referenced by `ptr`
+    /// could fit `layout`.
+    ///
+    /// Note that one cannot pass `CannotReallocInPlace` to the `oom`
+    /// method; clients are expected either to be able to recover from
+    /// `shrink_in_place` failures without aborting, or to fall back
+    /// on another reallocation method before resorting to an abort.
+    unsafe fn shrink_in_place_excess(&mut self,
+                              ptr: NonNull<Opaque>,
+                              layout: Layout,
+                              new_size: usize) -> Result<Excess, CannotReallocInPlace> {
+        let _ = ptr; // this default implementation doesn't care about the actual address.
+        debug_assert!(new_size <= layout.size);
+        let (l, u) = self.usable_size(&layout);
+        //                      layout.size() <= _u  [guaranteed by usable_size()]
+        // new_layout.size() <= layout.size()        [required by this method]
+        if l <= new_size {
+            return Ok(Excess(ptr, u));
+        } else {
+            return Err(CannotReallocInPlace);
```

Comment:
There are two issues that interact here:

* the user is not allowed to call `dealloc` with any size, but the size used must be in the range returned by `usable_size -> (l, u)`. Sizes in that range must not invoke undefined behavior when passed to `dealloc`. 

* the only requirement on the sizes `(l, u)` returned by `usable_size` is that `l <= u`. Therefore, a valid `usable_size` implementation is: `usable_size() -> (u, u).

Therefore, we can end here if the lower bound on the allocation size `l` returned by `usable_size -> (l, u)` is larger than `new_size`. 

If we were to return `Ok` in this case, the user would be allowed to call `dealloc` with `new_size`, but since that's out of the `[l, u]` range, that would trigger undefined behavior when calling `dealloc`. 

Does that make sense?

---

FWIW I think that the current API of `shrink_in_place` is a bit broken. For example, `malloc` does not care about the allocation size on deallocation, so a valid implementation of `usable_size` for malloc is to return `(0, size)`. This would allow implementing `shrink_in_place` / `shrink_in_place_excess` for `malloc` by just doing nothing and returning `Ok`, but I don't think that's a useful implementation of `shrink_in_place`. Even if it were to return `Ok` only if it was able to shrink the allocation, that might not be useful either. If I have a 1Gb allocation, and want to shrink it in place to 100 Mb, and `shrink_in_place` only shrinks it by 1 byte, then that's not useful. 

The wording that is problematic is that `shrink_in_place` is allowed to return more size than the one requested, which means that it can just return the original size and do nothing. 

For `grow_in_place` this wording makes sense, but it seems that it was c&p for `shrink_in_place`. So either `shrink_in_place` should always return the allocation size so that the user can check whether it shrunk the allocation enough, or it should take a requested size and an upper bound, and return error if it can't shrink the allocation under the upper bound.  

Also while I understand the rationale behind splitting `realloc_in_place` into `{grow,shrink}_in_place`, I am starting to doubt whether that was the right call. A single `realloc_in_place` method would reduce API surface and simplify things even if it adds a branch for allocators that can have completely different execution paths for growing and shrinking.

In any case all of this must be discussed in the RFC issue, and not here, but those are my thoughts on the matter.